index.test.js

```diff
@@ -47,17 +77,174 @@ it('should initialise IPFS API by accessing it directly when running inside of i
+      expect(store.selectIpfsReady()).toBe(true)
+      expect(store.selectIpfsIdentity()).toEqual(id)
+      done()
+    })
+
+    store.doInitIpfs()
+  })
+
+  it('Should connect via js-ipfs-api when window.ipfs present but disabled by options', (done) => {
+    const testIdentity = { hello: 'window.ipfs' }
+
+    global.ipfs = {
+      id: jest.fn().mockResolvedValue(testIdentity)
+    }
+
+    const store = composeBundlesRaw(
+      ipfsBundle({ tryWindow: false, defaultApiOpts: apiAddr })
+    )()
+
+    expect(store.selectIpfsReady()).toBe(false)
+    store.subscribeToSelectors(['selectIpfsReady'], () => {
+      expect(store.selectIpfsReady()).toBe(true)
+      expect(store.selectIpfsIdentity()).toEqual(id)
+      done()
+    })
+
+    store.doInitIpfs()
+  })
+
+  it('Should connect via js-ipfs-api when window.ipfs present but fails', (done) => {
+    global.ipfs = {
+      id: async () => {
+        throw new Error('STOP')
+      }
+    }
+
+    const store = composeBundlesRaw(
+      ipfsBundle({ defaultApiOpts: apiAddr })
+    )()
+
+    expect(store.selectIpfsReady()).toBe(false)
+    store.subscribeToSelectors(['selectIpfsReady'], () => {
+      expect(store.selectIpfsReady()).toBe(true)
+      expect(store.selectIpfsIdentity()).toEqual(id)
+      done()
+    })
 
-  expect(store.selectIpfsReady()).toBe(false)
-  store.subscribeToSelectors(['selectIpfsReady'], () => {
-    expect(store.selectIpfsReady()).toBe(true)
-    expect(store.selectIpfsIdentity()).toBe(testWebExtIdentity)
-    done()
+    store.doInitIpfs()
   })
-  store.doInitIpfs()
+})
+
+describe('js-ipfs', () => {
+  global.ipfs = undefined
+  global.browser = undefined
+
+  it('Should connect via js-ipfs', (done) => {
+    const store = composeBundlesRaw(
+      ipfsBundle({
+        tryJsIpfs: true,
+        Ipfs: require('ipfs')
+      })
+    )()
+
+    expect(store.selectIpfsReady()).toBe(false)
+    let first = true
+
+    store.subscribeToSelectors(['selectIpfsReady'], () => {
+      if (first) {
+        expect(store.selectIpfsReady()).toBe(true)
+        store.doStopIpfs()
+        first = false
+      } else {
+        expect(store.selectIpfsReady()).toBe(false)
+        done()
+      }
+    })
+
+    store.doInitIpfs()
+  }, longTimeout)
+
+  it('Should connect via js-ipfs instead of window.ipfs or js-ipfs-api', (done) => {
+    const testIdentity = { hello: 'window.ipfs' }
+
+    global.ipfs = {
+      id: jest.fn().mockResolvedValue(testIdentity)
+    }
+
+    const store = composeBundlesRaw(
+      ipfsBundle({
+        tryWindow: false,
+        tryApi: false,
+        tryJsIpfs: true,
+        Ipfs: require('ipfs')
+      })
+    )()
+
+    let first = true
+    expect(store.selectIpfsReady()).toBe(false)
+    store.subscribeToSelectors(['selectIpfsReady'], () => {
+      if (first) {
+        expect(store.selectIpfsReady()).toBe(true)
+        expect(store.selectIpfsIdentity()).not.toEqual(testIdentity)
```

Comment:
Testing that the identity is not equal to the test one doesn't tell us very much. It'd be useful to test for the value of `selectIpfsProvider` is set correctly in all these cases.